### PR TITLE
Phase2-sim27 Try to remove dependency of CLHEP in a few places and streamline the SteppingAction code for transporting to GeantV application

### PR DESCRIPTION
**Pull request by bsunanda**
Modify CaloGVHit class to include Event ID and CaloSteppingAction to make Geant4 relevant call in only 1 or 2 methods

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by kpedro88** [comment]
@bsunanda this is a very good step in the refactoring. Eventually we want to to make it so `CaloSteppingAction::update(const G4Step * aStep)` just extracts generic quantities from `aStep` and then passes them to an intermediate function like `update_(double energy, double charge, ...)`, in order to minimize the code duplication when adding the GeantV interface. Some of the complicated extractions of quantities (like geometry volumes) may go in their own helper functions.

File: Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc [hunk 1]
```diff
@@ -5,11 +5,10 @@
 
 #include "CondFormats/GeometryObjects/interface/HcalParameters.h"
 #include "Geometry/HcalCommonData/interface/HcalNumberingFromDDD.h"
+#include "DetectorDescription/DDCMS/interface/DDUnits.h"
```
Comment (cvuosalo) [inline Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc]:
```suggestion
#include "DetectorDescription/Core/interface/DDUnits.h"
```

Comment (cvuosalo) [reply Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc]: There are two versions of `DDUnits.h`. They will be consolidated and moved somewhere more central soon.

File: Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc [hunk 2]
```diff
@@ -5,11 +5,10 @@
...
 
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
 
-#include "CLHEP/Units/GlobalPhysicalConstants.h"
-#include "CLHEP/Units/GlobalSystemOfUnits.h"
 #include <iostream>
 
 //#define EDM_ML_DEBUG
```
Comment (cvuosalo) [inline Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc]:
```suggestion
using namespace dd;
using namespace dd::operators;

//#define EDM_ML_DEBUG
```

Comment (cvuosalo) [reply Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc]: Needed by the Core version of `DDUnits.h`.

File: Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc [hunk 3]
```diff
@@ -87,10 +86,10 @@ HcalNumberingFromDDD::HcalID HcalNumberingFromDDD::unitID(int det,
   int    ieta  = hcalConstants->getEta(det, lay, hetaR);
   std::pair<double,double> ficons = hcalConstants->getPhiCons(det, ieta);
 
-  int    nphi  = int((CLHEP::twopi+0.1*ficons.second)/ficons.second);
+  int    nphi  = int((2_pi+0.1*ficons.second)/ficons.second);
```
Comment (cvuosalo) [inline Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc]:
```suggestion
  int    nphi  = int((2._pi+0.1*ficons.second)/ficons.second);
```

File: Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc [hunk 4]
```diff
@@ -87,10 +86,10 @@ HcalNumberingFromDDD::HcalID HcalNumberingFromDDD::unitID(int det,
...
   int    zside = etaR>0 ? 1: 0;
   double hphi  = phi+ficons.first;
-  if (hphi < 0)    hphi += CLHEP::twopi;
+  if (hphi < 0)    hphi += (2_pi);
```
Comment (cvuosalo) [inline Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc]:
```suggestion
  if (hphi < 0)    hphi += (2._pi);
```

File: SimG4CMS/Calo/interface/CaloGVHit.h [hunk 5]
```diff
@@ -88,7 +93,8 @@ class CaloGVHitLess {
 class CaloGVHitEqual {
 public:
   bool operator()( const CaloGVHit* a, const CaloGVHit* b) {
-    return (a->getTrackID()     == b->getTrackID() && 
+    return (a->getEventID()     == b->getEventID() && 
+	    a->getTrackID()     == b->getTrackID() && 
 	    a->getUnitID()      == b->getUnitID()  && 
 	    a->getDepth()       == b->getDepth()   &&
 	    a->getTimeSliceID() == b->getTimeSliceID());
```
Comment (cvuosalo) [inline SimG4CMS/Calo/interface/CaloGVHit.h]: Only `inline` implementations are allowed in a header file: [4.3](https://github.com/cms-sw/cms-sw.github.io/blob/master/cms_coding_rules.md#4----technical-coding-rules-1)

Comment (makortel) [reply SimG4CMS/Calo/interface/CaloGVHit.h]: @cvuosalo Are you suggesting that there should be an `inline` keyword, or that the function definition should be moved to `.cc`?

Comment (bsunanda) [reply SimG4CMS/Calo/interface/CaloGVHit.h]: I am using inline keyword now

File: Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc (hunk 4 above)
Comment (bsunanda) [reply Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc]: done

File: Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc (hunk 3 above)
Comment (bsunanda) [reply Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc]: done

File: Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc (hunk 2 above)
Comment (bsunanda) [reply Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc]: Done

File: Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc (hunk 1 above)
Comment (bsunanda) [reply Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc]: Done

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: SimG4CMS/Calo/interface/CaloGVHit.h (hunk 5 above)
Comment (makortel) [reply SimG4CMS/Calo/interface/CaloGVHit.h]: The keyword is redundant in this context, we have tons of code without the keyword, and 4.3 does not explicitly require the use of the keyword.

Comment (kpedro88) [reply SimG4CMS/Calo/interface/CaloGVHit.h]: Yes, any function implemented inside a class definition is automatically inline.

Comment (cvuosalo) [reply SimG4CMS/Calo/interface/CaloGVHit.h]: I wanted to make sure all of the functions in CaloGVHit.h were intended as inline functions. Some of them are a bit long to be inline, but I guess they are fine like this.

**Comment by andrius-k** [comment]
+1

**Comment by cvuosalo** [comment]
+1

**Comment by ssekmen** [comment]
+1

File: Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc [hunk 6]
```diff
@@ -87,10 +88,10 @@ HcalNumberingFromDDD::HcalID HcalNumberingFromDDD::unitID(int det,
   int    ieta  = hcalConstants->getEta(det, lay, hetaR);
   std::pair<double,double> ficons = hcalConstants->getPhiCons(det, ieta);
 
-  int    nphi  = int((CLHEP::twopi+0.1*ficons.second)/ficons.second);
+  int    nphi  = int((2._pi+0.1*ficons.second)/ficons.second);
```
Comment (civanch) [inline Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc]: @bsunanda for me this "2._pi" looks terrible. If we introduce DD constants we should introduce namespaces like @VinInn pointed out in other PR and normal constants like DD::twopi

Comment (kpedro88) [reply Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc]: This follows several PRs from @cvuosalo (#24509, #25857) using the new user-defined literals feature from C++11: https://en.cppreference.com/w/cpp/language/user_literal

Comment (civanch) [reply Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc]: @cvuosalo , any constants assumed to be prepared at compilation, so here performance is not an issue. 

I just had a look into the list of literals and it seems to be incomplete: as minimum, ns, GeV and TeV seems to be useful.

**Comment by civanch** [comment]
+1

File: Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc (hunk 6 above)
Comment (cvuosalo) [reply Geometry/HcalCommonData/src/HcalNumberingFromDDD.cc]: @civanch: Thanks for the suggestions. I added them to the DDUnits.h issue: #25875.

**Comment by kpedro88** [comment]
+1

**Pull request merged**